File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        def traverse(curr_idx, curr_sum, curr_com):
            if curr_sum < target and curr_idx < len(nums):
                traverse(curr_idx, (curr_sum + nums[curr_idx]), curr_com + [nums[curr_idx]])
                traverse(curr_idx + 1, curr_sum, curr_com)
            else:
                if curr_sum == target:
                    res.append(curr_com[:])
                return
        traverse(0, 0, [])
        return res

File: Data Structures & Algorithms/binary-subarrays-with-sum/submission-0.cpp
class Solution {
    int subarraysUpToK(vector<int>& nums, int target){
        if (target < 0){
            return 0;
        }
        int sum = 0;
        int res = 0;
        int l = 0;
        int r = 0;
        while (r < nums.size()){
            sum += nums[r];
            while (sum > target){
                sum = sum - nums[l];
                l += 1;
            }
            res += (r - l + 1);
            r += 1;
        }
        return res;
    }
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return subarraysUpToK(nums, goal) - subarraysUpToK(nums, (goal - 1));
    }
};